<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Annotation fetching &mdash; hts-integrate - 0.4.2</title>
    
    <link rel="stylesheet" href="_static/agogo.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="hts-integrate - 0.4.2" href="index.html" />
    <link rel="next" title="Differential expression analysis" href="deseq-example.html" />
    <link rel="prev" title="BLAST" href="blast.html" /> 
  </head>
  <body>
    <div class="header-wrapper">
      <div class="header">
        <div class="headertitle"><a
          href="index.html">hts-integrate - 0.4.2</a></div>
        <div class="rel">
          <a href="blast.html" title="BLAST"
             accesskey="P">previous</a> |
          <a href="deseq-example.html" title="Differential expression analysis"
             accesskey="N">next</a> |
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a>
        </div>
       </div>
    </div>

    <div class="content-wrapper">
      <div class="content">
        <div class="document">
            
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="annotation-fetching">
<h1>Annotation fetching<a class="headerlink" href="#annotation-fetching" title="Permalink to this headline">¶</a></h1>
<p>A very common bioinformatics task is that of obtaining current annotations for a given gene or protein.  There is a great deal of flexibility in how annotations may be obtained from the database and some strategies are more or less appropriate for a specific application.  First, however it is important that you understand how the <a class="reference external" href="http://geneontology.org">Gene Ontology (GO)</a> information is stored in the <a class="reference internal" href="database.html"><em>database</em></a>.</p>
<div class="section" id="fetching-annotations-with-a-list-of-ncbi-or-uniprot-identifiers">
<h2>Fetching annotations with a list of NCBI or UniProt identifiers<a class="headerlink" href="#fetching-annotations-with-a-list-of-ncbi-or-uniprot-identifiers" title="Permalink to this headline">¶</a></h2>
<p>First, we need to connect to the database.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">htsint.database</span> <span class="kn">import</span> <span class="n">fetch_annotations</span><span class="p">,</span><span class="n">db_connect</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">session</span><span class="p">,</span> <span class="n">engine</span> <span class="o">=</span> <span class="n">db_connect</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">conn</span> <span class="o">=</span> <span class="n">engine</span><span class="o">.</span><span class="n">connect</span><span class="p">()</span>
</pre></div>
</div>
<p>The <tt class="docutils literal"><span class="pre">conn</span></tt> or connection is for usage with the <a class="reference external" href="http://docs.sqlalchemy.org/en/rel_0_9/core/">SQLAlchemy core API</a> and is not necessary for basic annotation fetching.  If you have a list of NCBI gene identifiers or UniProt identifiers (UniprotAc) then the easiest way to get the annotations is to</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">annotations</span> <span class="o">=</span> <span class="n">fetch_annotations</span><span class="p">([</span><span class="s">&#39;31251&#39;</span><span class="p">],</span><span class="n">engine</span><span class="p">,</span><span class="n">idType</span><span class="o">=</span><span class="s">&#39;ncbi&#39;</span><span class="p">,</span><span class="n">useIea</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span><span class="n">aspect</span><span class="o">=</span><span class="s">&#39;biological_process&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">annotations</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">annotations</span><span class="p">[</span><span class="s">&#39;31251&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
<span class="go">(&#39;GO:0042752&#39;, &#39;regulation of circadian rhythm&#39;)</span>
</pre></div>
</div>
<p>Simply use <cite>idType=&#8217;uniprot&#8217;</cite> if the ids are from that namespace.  As you can see the results are contained in a dictionary with a key for each NCBI identifier.  The results associated with each identifier are a unique list of tuples each containing the GO identifier and GO name. So if you wanted only the GO names simply use a list comprehension.</p>
<blockquote>
<div><div class="highlight-python"><div class="highlight"><pre><span class="k">print</span><span class="p">([</span><span class="n">a</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">annotations</span><span class="p">[</span><span class="s">&#39;31251&#39;</span><span class="p">]])</span>
</pre></div>
</div>
</div></blockquote>
<p>A results list can be quickly sorted with:</p>
<blockquote>
<div><div class="highlight-python"><div class="highlight"><pre><span class="n">sortby_inplace</span><span class="p">(</span><span class="n">annotations</span><span class="p">[</span><span class="s">&#39;31251&#39;</span><span class="p">],</span> <span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div></blockquote>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">For large lists it may be faster to search for all annotations using <tt class="docutils literal"><span class="pre">fetch_taxa_annotations</span></tt> (see below) then to trim the list</p>
</div>
</div>
<div class="section" id="fetching-directly-with-sqlalchemy">
<h2>Fetching directly with SQLAlchemy<a class="headerlink" href="#fetching-directly-with-sqlalchemy" title="Permalink to this headline">¶</a></h2>
<p>There are two API&#8217;s from SQLAlchemy the ORM and the Core.</p>
<div class="section" id="using-the-orm">
<h3>Using the ORM<a class="headerlink" href="#using-the-orm" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">subq1</span> <span class="o">=</span> <span class="n">session</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">GoTerm</span><span class="p">)</span><span class="o">.</span>\
<span class="go">            filter(GoTerm.aspect==aspect).\</span>
<span class="go">            subquery()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">subq2</span> <span class="o">=</span> <span class="n">session</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">GoAnnotation</span><span class="p">)</span><span class="o">.</span>\
<span class="go">          filter(GoAnnotation.gene_id==gquery[&#39;id&#39;]).\</span>
<span class="go">          subquery()</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">q</span> <span class="o">=</span> <span class="n">session</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">subq1</span><span class="o">.</span><span class="n">c</span><span class="o">.</span><span class="n">name</span><span class="p">)</span><span class="o">.</span>\
<span class="go">        filter(subq1.c.id==subq2.c.go_term_id).all()</span>
</pre></div>
</div>
</div>
<div class="section" id="using-the-core">
<h3>Using the Core<a class="headerlink" href="#using-the-core" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">select</span><span class="p">([</span><span class="n">GoTerm</span><span class="o">.</span><span class="n">name</span><span class="p">],</span><span class="n">GoAnnotation</span><span class="o">.</span><span class="n">go_term_id</span><span class="o">==</span><span class="n">GoTerm</span><span class="o">.</span><span class="n">id</span><span class="p">)</span><span class="o">.</span>\
<span class="go">      where(GoAnnotation.gene_id == gquery[&#39;id&#39;]).\</span>
<span class="go">      where(GoTerm.aspect==aspect)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">_results</span> <span class="o">=</span> <span class="n">conn</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span> <span class="n">results</span> <span class="o">=</span> <span class="p">[</span><span class="nb">tuple</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span><span class="n">r</span><span class="p">))</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">_results</span><span class="o">.</span><span class="n">fetchall</span><span class="p">()]</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="fetching-annotations-by-taxa">
<h2>Fetching annotations by taxa<a class="headerlink" href="#fetching-annotations-by-taxa" title="Permalink to this headline">¶</a></h2>
<p>The syntax is similar to individual gene/protein identifiers.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">htsint.database</span> <span class="kn">import</span> <span class="n">fetch_taxa_annotations</span><span class="p">,</span><span class="n">db_connect</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">session</span><span class="p">,</span> <span class="n">engine</span> <span class="o">=</span> <span class="n">db_connect</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">conn</span> <span class="o">=</span> <span class="n">engine</span><span class="o">.</span><span class="n">connect</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">geneAnnots</span><span class="p">,</span><span class="n">uniprotAnnots</span> <span class="o">=</span> <span class="n">fetch_taxa_annotations</span><span class="p">([</span><span class="s">&#39;7091&#39;</span><span class="p">],</span><span class="n">engine</span><span class="p">,</span><span class="n">useIea</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span><span class="n">verbose</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">uniprotAnnots</span><span class="p">[</span><span class="s">&#39;Q9NL89&#39;</span><span class="p">])</span>
<span class="go">[&#39;GO:0045088&#39;]</span>
</pre></div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
        </div>
        <div class="sidebar">
          <h3>Table Of Contents</h3>
          <ul>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="database.html">Database</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="database-cookbook.html">Database cookbook</a></li>
<li class="toctree-l1"><a class="reference internal" href="blast.html">Using BLAST</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="">Gene Ontology annotation fetching</a></li>
<li class="toctree-l1"><a class="reference internal" href="deseq-example.html">Gene level differential expression</a></li>
<li class="toctree-l1"><a class="reference internal" href="gsa-example.html">Gene set analysis example</a></li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References</a></li>
</ul>

          <h3 style="margin-top: 1.5em;">Search</h3>
          <form class="search" action="search.html" method="get">
            <input type="text" name="q" />
            <input type="submit" value="Go" />
            <input type="hidden" name="check_keywords" value="yes" />
            <input type="hidden" name="area" value="default" />
          </form>
          <p class="searchtip" style="font-size: 90%">
            Enter search terms or a module, class or function name.
          </p>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

    <div class="footer-wrapper">
      <div class="footer">
        <div class="left">
          <a href="blast.html" title="BLAST"
             >previous</a> |
          <a href="deseq-example.html" title="Differential expression analysis"
             >next</a> |
          <a href="genindex.html" title="General Index"
             >index</a>
            <br/>
            <a href="_sources/annotation-fetching.txt"
               rel="nofollow">Show Source</a>
        </div>

        <div class="right">
          
    <div class="footer">
        &copy; Copyright 2015, A. Richards.
      Last updated on Apr 05, 2015.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.2.
    </div>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

  </body>
</html>