<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>BLAST &mdash; hts-integrate - 0.4.1</title>
    
    <link rel="stylesheet" href="_static/agogo.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="hts-integrate - 0.4.1" href="index.html" />
    <link rel="next" title="Annotation fetching" href="annotation-fetching.html" />
    <link rel="prev" title="Common database tasks" href="database-cookbook.html" /> 
  </head>
  <body>
    <div class="header-wrapper">
      <div class="header">
        <div class="headertitle"><a
          href="index.html">hts-integrate - 0.4.1</a></div>
        <div class="rel">
          <a href="database-cookbook.html" title="Common database tasks"
             accesskey="P">previous</a> |
          <a href="annotation-fetching.html" title="Annotation fetching"
             accesskey="N">next</a> |
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a>
        </div>
       </div>
    </div>

    <div class="content-wrapper">
      <div class="content">
        <div class="document">
            
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="blast">
<h1>BLAST<a class="headerlink" href="#blast" title="Permalink to this headline">¶</a></h1>
<p>BLAST or Basic Local Alignment Search Tool is a widely used tool in computational biology.  A general introduction is beyond the scope of these documents, but more information can be found by visiting the following sites.</p>
<blockquote>
<div><ul class="simple">
<li><a class="reference external" href="http://blast.ncbi.nlm.nih.gov/Blast.cgi">NCBI BLAST page</a></li>
<li><a class="reference external" href="http://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs&amp;DOC_TYPE=FAQ">NCBI BLAST FAQs</a></li>
<li><a class="reference external" href="http://www.ncbi.nlm.nih.gov/books/NBK1734/">BLAST QuickStart</a></li>
</ul>
</div></blockquote>
<div class="section" id="install-and-setup">
<h2>Install and setup<a class="headerlink" href="#install-and-setup" title="Permalink to this headline">¶</a></h2>
<p>Local BLAST can be run through <tt class="docutils literal"><span class="pre">htsint</span></tt> or through the command line. BLAST results help integrate data across taxa through comparative genomics.  For local BLAST functionality, install the Suite of <a class="reference external" href="http://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download">BLAST+ tools</a>.  Under Ubuntu you can simply use:</p>
<blockquote>
<div><div class="highlight-bash"><div class="highlight"><pre>~<span class="nv">$ </span>sudo apt-get install ncbi-blast+
</pre></div>
</div>
</div></blockquote>
</div>
<div class="section" id="id1">
<h2>BLAST<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h2>
<p>If you have successfully setup the <tt class="docutils literal"><span class="pre">htsint</span></tt> <a class="reference internal" href="database.html"><em>database</em></a> then the <a class="reference external" href="http://web.expasy.org/docs/swiss-prot_guideline.html">SwissProt</a> database has been downloaded into your data directory.  Any other database or organism that you wish to BLAST against may be downloaded to this directory.  For example, to download the protein sequences of <em>Anolis carolinensis</em> and make the appropriate BLAST files use the following.</p>
<blockquote>
<div><div class="highlight-bash"><div class="highlight"><pre>~<span class="nv">$ </span><span class="nb">cd</span> /your/data/directory
~<span class="nv">$ </span>wget ftp://ftp.ncbi.nlm.nih.gov/genomes/Anolis_carolinensis/protein/protein.fa.gz
~<span class="nv">$ </span>gunzip -c protein.fa.gz &gt; Anolis_carolinensis.fasta
~<span class="nv">$ </span>makeblastdb -in Anolis_carolinensis.fasta -dbtype <span class="s1">&#39;prot&#39;</span> -out Anolis_carolinensis
</pre></div>
</div>
</div></blockquote>
<p>To run BLAST from within <tt class="docutils literal"><span class="pre">htsint</span></tt> use the Blast class.  For convenience <tt class="docutils literal"><span class="pre">htsint</span></tt> uses the sequence classes in <a class="reference external" href="http://biopython.org/wiki/Main_Page">BioPython</a> to handle basic operations.  Here is the FASTA file used in this example.</p>
<blockquote>
<div><ul class="simple">
<li><a class="reference download internal" href="_downloads/xtrop.fasta"><tt class="xref download docutils literal"><span class="pre">xtrop.fasta</span></tt></a></li>
</ul>
</div></blockquote>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">htsint.blast</span> <span class="kn">import</span> <span class="n">Blast</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">blast</span> <span class="o">=</span> <span class="n">Blast</span><span class="p">(</span><span class="s">&#39;xtrop.fasta&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">targetDB</span> <span class="o">=</span> <span class="s">&quot;uniprot_sprot.fasta.db&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">blastFile</span> <span class="o">=</span> <span class="n">blast</span><span class="o">.</span><span class="n">run_blast</span><span class="p">(</span><span class="n">targetDB</span><span class="p">,</span><span class="n">evalue</span><span class="o">=</span><span class="mf">0.0001</span><span class="p">,</span><span class="n">start</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="n">stop</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span><span class="n">cmd</span><span class="o">=</span><span class="s">&#39;blastx&#39;</span><span class="p">)</span>
<span class="go">Running blast</span>
<span class="go">Total run time: 00:00:07</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">blastFile</span><span class="p">)</span>
<span class="go">./xtrop-0-2.xml</span>
</pre></div>
</div>
<p>In this example the output file contains only two sequences as constrained by start and stop. Indexing in Python begin with zero.  If you type <tt class="docutils literal"><span class="pre">range(start,stop)</span></tt> it can help you understand exactly which indexes are implied.  The start and stop options are optional, but they are convenient when moving BLAST into a cluster environment.  If stop is larger than the number of sequences it will still stop at the last one in the file.  The outfile name is automatically created with numbers indicating the first and last sequence of the input FASTA file.</p>
<p>This produces the XML output file that was specified and a log file.  To do BLAST the entire file from the command line use you may use:</p>
<blockquote>
<div><div class="highlight-bash"><div class="highlight"><pre>~<span class="nv">$ </span><span class="nb">export </span><span class="nv">BLASTDB</span><span class="o">=</span><span class="s1">&#39;/usr/local/share/htsint/:$BLASTDB&#39;</span>
~<span class="nv">$ </span>blastx -query xtrop.fasta -db uniprot_sprot.fasta.db -out xtrop.xml -evalue 0.0001 -num_threads 8 -outfmt 5
</pre></div>
</div>
</div></blockquote>
<p>Because BLAST results are fairly large by default we parse them into a more reasonable size.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">htsint.blast</span> <span class="kn">import</span> <span class="n">ParseBlast</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">parser</span> <span class="o">=</span> <span class="n">ParseBlast</span><span class="p">(</span><span class="n">blastFile</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">parsedFile</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>
<span class="go">total hits: 2</span>
<span class="go">sequences with at least one match : 2</span>
<span class="go">sequences without any matches: 0</span>
<span class="go">complete.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">parsedFile</span><span class="p">)</span>
<span class="go">./xtrop-0-2_parsed.csv</span>
</pre></div>
</div>
<div class="section" id="blast-in-a-cluster-environment">
<h3>BLAST in a cluster environment<a class="headerlink" href="#blast-in-a-cluster-environment" title="Permalink to this headline">¶</a></h3>
<p>Wrappers for the Blast and ParseBlast classes have been created for grid engine cluster environments.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">htsint.blast</span> <span class="kn">import</span> <span class="n">ParallelBlast</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">parBlast</span> <span class="o">=</span> <span class="n">ParallelBlast</span><span class="p">(</span><span class="n">queryFile</span><span class="o">.</span><span class="n">fasta</span><span class="p">,</span><span class="s">&#39;uniprot_sprot.fasta.db&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">parBlast</span><span class="o">.</span><span class="n">evalue</span> <span class="o">=</span> <span class="mf">0.0001</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">chunks</span> <span class="o">=</span> <span class="mi">30</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">parBlast</span><span class="o">.</span><span class="n">create_scripts</span><span class="p">(</span><span class="n">chunks</span><span class="p">,</span><span class="s">&quot;myemail@somewhere.edu&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">parBlast</span><span class="o">.</span><span class="n">submit</span><span class="p">()</span>
</pre></div>
</div>
<p>The individual scripts will be created in a folder called <tt class="docutils literal"><span class="pre">cluster</span></tt> and chunks are the number of jobs you wish to break the FASTA file into.  To parse the results use:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">htsint.blast</span> <span class="kn">import</span> <span class="n">ParseParallelBlast</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">parser</span> <span class="o">=</span> <span class="n">ParseParallelBlast</span><span class="p">(</span><span class="n">queryFilePath</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">chunks</span> <span class="o">=</span> <span class="mi">30</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">parser</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">chunks</span><span class="p">)</span>
</pre></div>
</div>
<p>This will create a parsed and concatenated file in the <tt class="docutils literal"><span class="pre">cluster</span></tt> directory.</p>
</div>
</div>
<div class="section" id="blastmapper">
<h2>BlastMapper<a class="headerlink" href="#blastmapper" title="Permalink to this headline">¶</a></h2>
<p>BlastMapper is a class that htsint uses to interact with BLAST results.  BlastMapper takes as input the parsed results.  A method called <tt class="docutils literal"><span class="pre">create_summarized</span></tt> will take the parsed results and try to find matching UniProt and NCBI identifiers from the local database.  This summary can take time so it is saved as a separate file.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">htsint.blast</span> <span class="kn">import</span> <span class="n">BlastMapper</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bm</span> <span class="o">=</span> <span class="n">BlastMapper</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">summaryFile</span> <span class="o">=</span> <span class="n">bm</span><span class="o">.</span><span class="n">create_summarized</span><span class="p">(</span><span class="n">parsedFile</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">summaryFile</span><span class="p">)</span>
<span class="go">./xtrop-0-2_parsed_summary.csv</span>
</pre></div>
</div>
<p>BLAST results will most commonly be used in the summarized form so there is some flexibility in how the results may be loaded.  The object of interest is the dictionary that gets returned.  The keys will always correspond to the query identifiers.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">bmap</span> <span class="o">=</span> <span class="n">bm</span><span class="o">.</span><span class="n">load_summary</span><span class="p">(</span><span class="n">summaryFile</span><span class="p">,</span><span class="n">best</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span><span class="n">evalue</span><span class="o">=</span><span class="mf">0.0001</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">bmap</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
<span class="go">[&#39;Xetro.K04500.1&#39;, &#39;Xetro.K04761.1&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">bmap</span><span class="p">[</span><span class="s">&#39;Xetro.K04500.1&#39;</span><span class="p">])</span>
<span class="go">(&#39;CJ088_HUMAN&#39;, &#39;80007&#39;, &#39;Homo sapiens&#39;, &#39;9606&#39;, 3.528e-35)</span>
</pre></div>
</div>
<p>Because the argument <tt class="docutils literal"><span class="pre">best</span></tt> was set to True only the best matching result for each query is returned.  A result is a tuple of the form</p>
<blockquote>
<div><div class="highlight-none"><div class="highlight"><pre>(database_id, ncbi_gene_id, species_name, species_id, e-value)
</pre></div>
</div>
</div></blockquote>
<p>If <tt class="docutils literal"><span class="pre">best</span></tt> is set to False then all hits passing the threshold e-value (default=0.0001) are returned.  The returned object changes from a tuple to a list of tuples.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">bmap</span> <span class="o">=</span> <span class="n">bm</span><span class="o">.</span><span class="n">load_summary</span><span class="p">(</span><span class="n">summaryFile</span><span class="p">,</span><span class="n">best</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span><span class="n">evalue</span><span class="o">=</span><span class="mf">0.0001</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">hit</span> <span class="ow">in</span> <span class="n">bmap</span><span class="p">[</span><span class="s">&#39;Xetro.K04500.1&#39;</span><span class="p">]:</span>
<span class="go">        print(hit)</span>
<span class="go">(&#39;CJ088_HUMAN&#39;, &#39;80007&#39;, &#39;Homo sapiens&#39;, &#39;9606&#39;, 3.528e-35)</span>
<span class="go">(&#39;CJ088_MOUSE&#39;, &#39;68277&#39;, &#39;Mus musculus&#39;, &#39;10090&#39;, 1.11258e-31)</span>
<span class="go">(&#39;CJ088_RAT&#39;, &#39;309029&#39;, &#39;Rattus norvegicus&#39;, &#39;10116&#39;, 9.44503e-30)</span>
<span class="go">(&#39;YJ016_HUMAN&#39;, &#39;-&#39;, &#39;Homo sapiens&#39;, &#39;9606&#39;, 7.09757e-09)</span>
</pre></div>
</div>
<p>We can also restrict the results by taxa.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">bmap</span> <span class="o">=</span> <span class="n">bm</span><span class="o">.</span><span class="n">load_summary</span><span class="p">(</span><span class="n">summaryFile</span><span class="p">,</span><span class="n">best</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span><span class="n">evalue</span><span class="o">=</span><span class="mf">0.0001</span><span class="p">,</span><span class="n">taxaList</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;10090&#39;</span><span class="p">,</span><span class="s">&#39;9606&#39;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">hit</span> <span class="ow">in</span> <span class="n">bmap</span><span class="p">[</span><span class="s">&#39;Xetro.K04500.1&#39;</span><span class="p">]:</span>
<span class="go">        print(hit)</span>
<span class="go">(&#39;CJ088_HUMAN&#39;, &#39;80007&#39;, &#39;Homo sapiens&#39;, &#39;9606&#39;, 3.528e-35)</span>
<span class="go">(&#39;CJ088_MOUSE&#39;, &#39;68277&#39;, &#39;Mus musculus&#39;, &#39;10090&#39;, 1.11258e-31)</span>
<span class="go">(&#39;YJ016_HUMAN&#39;, &#39;-&#39;, &#39;Homo sapiens&#39;, &#39;9606&#39;, 7.09757e-09)</span>
</pre></div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
        </div>
        <div class="sidebar">
          <h3>Table Of Contents</h3>
          <ul>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="database.html">Database</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="database-cookbook.html">Database cookbook</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="">Using BLAST</a></li>
<li class="toctree-l1"><a class="reference internal" href="annotation-fetching.html">Gene Ontology annotation fetching</a></li>
<li class="toctree-l1"><a class="reference internal" href="deseq-example.html">Gene level differential expression</a></li>
<li class="toctree-l1"><a class="reference internal" href="gsa-example.html">Gene set analysis example</a></li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References</a></li>
</ul>

          <h3 style="margin-top: 1.5em;">Search</h3>
          <form class="search" action="search.html" method="get">
            <input type="text" name="q" />
            <input type="submit" value="Go" />
            <input type="hidden" name="check_keywords" value="yes" />
            <input type="hidden" name="area" value="default" />
          </form>
          <p class="searchtip" style="font-size: 90%">
            Enter search terms or a module, class or function name.
          </p>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

    <div class="footer-wrapper">
      <div class="footer">
        <div class="left">
          <a href="database-cookbook.html" title="Common database tasks"
             >previous</a> |
          <a href="annotation-fetching.html" title="Annotation fetching"
             >next</a> |
          <a href="genindex.html" title="General Index"
             >index</a>
            <br/>
            <a href="_sources/blast.txt"
               rel="nofollow">Show Source</a>
        </div>

        <div class="right">
          
    <div class="footer">
        &copy; Copyright 2015, A. Richards.
      Last updated on Mar 26, 2015.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.2.
    </div>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

  </body>
</html>