<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Database &mdash; htsint - 0.5.2</title>
    
    <link rel="stylesheet" href="_static/agogo.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="htsint - 0.5.2" href="index.html" />
    <link rel="next" title="Common database tasks" href="database-cookbook.html" />
    <link rel="prev" title="Installation" href="install.html" /> 
  </head>
  <body>
    <div class="header-wrapper">
      <div class="header">
        <div class="headertitle"><a
          href="index.html">htsint - 0.5.2</a></div>
        <div class="rel">
          <a href="install.html" title="Installation"
             accesskey="P">previous</a> |
          <a href="database-cookbook.html" title="Common database tasks"
             accesskey="N">next</a> |
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a>
        </div>
       </div>
    </div>

    <div class="content-wrapper">
      <div class="content">
        <div class="document">
            
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="database">
<h1>Database<a class="headerlink" href="#database" title="Permalink to this headline">¶</a></h1>
<p>The schema for the database.</p>
<div class="align-center figure">
<a class="reference internal image-reference" href="_images/dbschema.png"><img alt="database schema" src="_images/dbschema.png" style="width: 699.93px; height: 393.03px;" /></a>
</div>
<p>The main uses of the database are documented here:</p>
<blockquote>
<div><ul class="simple">
<li><a class="reference internal" href="database-cookbook.html"><em>database cookbook</em></a></li>
<li><a class="reference internal" href="annotation-fetching.html"><em>annotation fetching</em></a></li>
</ul>
</div></blockquote>
<div class="section" id="setting-up-your-database">
<h2>Setting up your database<a class="headerlink" href="#setting-up-your-database" title="Permalink to this headline">¶</a></h2>
<p>There are 4 step that you need to carry out in order to have a working database.</p>
<ol class="arabic simple">
<li>Install <a class="reference external" href="http://www.postgresql.org">PostgreSQL</a> and create an empty database</li>
<li>Modify the <cite>htsint</cite> config file</li>
<li>Fetch all necessary data files</li>
<li>Populate the datbase</li>
</ol>
<div class="section" id="postgresql-setup">
<h3>(1) PostgreSQL setup<a class="headerlink" href="#postgresql-setup" title="Permalink to this headline">¶</a></h3>
<p>Install <a class="reference external" href="http://www.postgresql.org">PostgreSQL</a> on you system.  For example on Ubuntu</p>
<blockquote>
<div><div class="highlight-bash"><div class="highlight"><pre>~<span class="nv">$ </span>sudo apt-get install postgresql
</pre></div>
</div>
</div></blockquote>
<p>Then you need to create a database.  Here we do this with a database named <cite>htsintegrate</cite>, a user named <cite>dbuser</cite> and a password <cite>somepassword</cite>.  You should change these parameters to fit your system.</p>
<blockquote>
<div><div class="highlight-bash"><div class="highlight"><pre>~<span class="nv">$ </span>sudo su - postgres
~<span class="nv">$ </span>psql -U postgres
CREATE USER dbuser WITH ENCRYPTED PASSWORD <span class="s1">&#39;somepassword&#39;</span>;
CREATE DATABASE htsintegrate WITH OWNER dbuser;
<span class="se">\q</span>
</pre></div>
</div>
</div></blockquote>
<p>Visit the <a class="reference external" href="http://www.postgresql.org/docs">PostgreSQL documentation</a> to learn more.</p>
</div>
<div class="section" id="modify-the-config-file">
<h3>(2) Modify the config file<a class="headerlink" href="#modify-the-config-file" title="Permalink to this headline">¶</a></h3>
<p>Set the database name, user, password and port</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">htsint</span> <span class="kn">import</span> <span class="n">Configure</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">config</span> <span class="o">=</span> <span class="n">Configure</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">config</span><span class="o">.</span><span class="n">log</span><span class="p">[</span><span class="s">&#39;dbuser&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;dbuser&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">config</span><span class="o">.</span><span class="n">log</span><span class="p">[</span><span class="s">&#39;dbname&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;htsintegrate&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">config</span><span class="o">.</span><span class="n">log</span><span class="p">[</span><span class="s">&#39;dbpass&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;somepassword&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">config</span><span class="o">.</span><span class="n">save</span><span class="p">()</span>
</pre></div>
</div>
<p>The other variables than can be modified are:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">config</span><span class="o">.</span><span class="n">log</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
<span class="go">[&#39;dbport&#39;, &#39;taxa&#39;, &#39;dbhost&#39;, &#39;dbpass&#39;, &#39;dbuser&#39;, &#39;blast&#39;, &#39;data&#39;, &#39;dbname&#39;]</span>
</pre></div>
</div>
<p>Check that all taxa that you would like annotation information for are present.  You can print the default taxa with</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">config</span><span class="o">.</span><span class="n">log</span><span class="p">[</span><span class="s">&#39;taxa&#39;</span><span class="p">])</span>
</pre></div>
</div>
<p>The default taxa are:</p>
<blockquote>
<div><div class="highlight-none"><div class="highlight"><pre>3702, Arabidopsis thaliana
4932, Saccharomyces cerevisiae
5476, Candida albicans
7227, Drosophila melanogaster
7955, Danio rerio
8355, Xenopus laevis
8364, Xenopus (Silurana) tropicalis
9031, Gallus gallus
9606, Homo sapiens
10090, Mus musculus
10566, Human papillomavirus
10116, Rattus norvegicus
28377, Anolis carolinensis
</pre></div>
</div>
</div></blockquote>
<p>If you would like to add, for example,  <em>Solanum lycopersicum</em> then use typical list syntax.  You need to save any changes you make.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">config</span><span class="o">.</span><span class="n">log</span><span class="p">[</span><span class="s">&#39;taxa&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s">&#39;4081&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">config</span><span class="o">.</span><span class="n">save</span><span class="p">()</span>
</pre></div>
</div>
<p>There is a good chance you will want to specify the directory where all the downloaded data is stored.  This can be done with any valid full path.</p>
<p>The default is</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">config</span><span class="o">.</span><span class="n">log</span><span class="p">[</span><span class="s">&#39;data&#39;</span><span class="p">])</span>
<span class="go">&#39;/usr/local/share/htsint&#39;</span>
</pre></div>
</div>
<p>Alternatively, it is possible to edit this file directly.  To locate the directory where it is stored you may type the following.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">os</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">expanduser</span><span class="p">(</span><span class="s">&#39;~&#39;</span><span class="p">),</span><span class="s">&quot;.htsint&quot;</span><span class="p">)</span>
<span class="go">&#39;/home/adam/.htsint&#39;</span>
</pre></div>
</div>
<p>The dbport (default &#8216;5432&#8217;) and dbhost (default &#8216;localhost&#8217;) may also be configured.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last"><tt class="docutils literal"><span class="pre">htsint</span></tt> will only populate annotation information for taxa in the <em>taxa</em> variable so make sure all species are present <strong>before</strong> database population.</p>
</div>
</div>
<div class="section" id="fetch-the-necessary-data-files">
<h3>(3) Fetch the necessary data files<a class="headerlink" href="#fetch-the-necessary-data-files" title="Permalink to this headline">¶</a></h3>
<p>The rest of the process is automated assuming you have created your PostgreSQL database and modified you config file.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">htsint.database</span> <span class="kn">import</span> <span class="n">DatabaseFetch</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fetch</span> <span class="o">=</span> <span class="n">DatabaseFetch</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fetch</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>
</pre></div>
</div>
<p>This class only currently works under Linux/OSX operating systems.  For other systems the following files could be downloaded by hand and placed in the &#8216;data&#8217; directory.</p>
<blockquote>
<div><ul class="simple">
<li><a class="reference external" href="ftp://ftp.geneontology.org/pub/go/ontology/go.obo">go.obo</a></li>
<li><a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdump.tar.gz">taxdump.tar.gz</a></li>
<li><a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz">gene_info.gz</a></li>
<li><a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene2go.gz">gene2go.gz</a></li>
<li><a class="reference external" href="ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz">gene_association.goa_uniprot.gz</a></li>
<li><a class="reference external" href="ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/idmapping.dat.gz">idmapping.dat.gz</a></li>
<li><a class="reference external" href="ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz">uniprot_sprot.fasta.gz</a></li>
</ul>
</div></blockquote>
<p>The fetching can take several hours depending on the speed of your connection.  The compressed files total less than 15GB, but be aware that the uncompressed versions will take up over 100GB of space.  If space is an issue all files may be erased except <tt class="docutils literal"><span class="pre">uniprot_sprot.fasta.*</span></tt> and <tt class="docutils literal"><span class="pre">go.obo</span></tt> as the former is used for BLAST and the latter is not stored directly in the database and is used as part of most analysis pipelines.</p>
<p>A logfile is produced and stored in your data directory.</p>
</div>
<div class="section" id="populate-the-database">
<h3>(4) Populate the database<a class="headerlink" href="#populate-the-database" title="Permalink to this headline">¶</a></h3>
<p>Finally, the database can be populated with the following class.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">htsint.database</span> <span class="kn">import</span> <span class="n">DatabaseCreate</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">db</span> <span class="o">=</span> <span class="n">DatabaseCreate</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">db</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>
</pre></div>
</div>
<p>A typical database will take a little over an hour to populate.  A logfile is produced and stored in your data directory.</p>
<p>A summary can be produced at any time using which will produce a similar output.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">htsint.database</span> <span class="kn">import</span> <span class="n">print_db_summary</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">print_db_summary</span><span class="p">()</span>
</pre></div>
</div>
<blockquote>
<div><div class="highlight-none"><div class="highlight"><pre>DATABASE - htsintegrate - SUMMARY
There are 1262260 entries in the taxa table
There are 681732 entries in the genes table
There are 777608 entries in the uniprot table
There are 42627 entries in the go_terms table
There are 7463568 entries in the go_annotations table
</pre></div>
</div>
</div></blockquote>
</div>
</div>
<div class="section" id="additional-notes">
<h2>Additional Notes<a class="headerlink" href="#additional-notes" title="Permalink to this headline">¶</a></h2>
<div class="section" id="what-exactly-is-stored-in-the-database">
<h3>What exactly is stored in the database?<a class="headerlink" href="#what-exactly-is-stored-in-the-database" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><ul class="simple">
<li>All taxa from <a class="reference external" href="http://www.ncbi.nlm.nih.gov/taxonomy">NCBI taxonomy</a></li>
<li>Gene, UniProt and GO annotation information for only the specified taxa</li>
<li>All information about GO terms</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="database-updating">
<h3>Database updating<a class="headerlink" href="#database-updating" title="Permalink to this headline">¶</a></h3>
<p>Because of the challenges that can arise through naming conflicts when updating NCBI and UniProt data it is recommended that you run the fetch and create steps again, which will create a clean updated version.  The fetch step will check if a current file is the newest and only download a new one if necessary.</p>
</div>
<div class="section" id="database-portability">
<h3>Database portability<a class="headerlink" href="#database-portability" title="Permalink to this headline">¶</a></h3>
<p>You may want to copy your database to another computer instead of waiting for the db to populate.  This can be done as follows.</p>
<p>To create a file that may be transferred to another computer</p>
<blockquote>
<div><div class="highlight-bash"><div class="highlight"><pre>~<span class="nv">$ </span>pg_dump -h localhost -U dbuser dbname &gt; htsint.sql
</pre></div>
</div>
</div></blockquote>
<p>To add the database to another server</p>
<blockquote>
<div><div class="highlight-bash"><div class="highlight"><pre>~<span class="nv">$ </span>sudo su - postgres
~<span class="nv">$ </span>psql -U postgres
CREATE USER dbuser WITH ENCRYPTED PASSWORD <span class="s1">&#39;somepassword&#39;</span>;
CREATE DATABASE newdbname WITH OWNER dbuser;
<span class="se">\q</span>
~<span class="nv">$ </span>psql newdbname &lt; htsint.sql
</pre></div>
</div>
</div></blockquote>
<p>See the <a class="reference internal" href="database-cookbook.html"><em>database cookbook</em></a> for more information on getting started with the database.</p>
</div>
</div>
</div>


          </div>
        </div>
      </div>
        </div>
        <div class="sidebar">
          <h3>Table Of Contents</h3>
          <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="">Database</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="database-cookbook.html">Database cookbook</a></li>
<li class="toctree-l1"><a class="reference internal" href="blast.html">Using BLAST</a></li>
<li class="toctree-l1"><a class="reference internal" href="annotation-fetching.html">Gene Ontology annotation fetching</a></li>
<li class="toctree-l1"><a class="reference internal" href="deseq-example.html">Gene level differential expression</a></li>
<li class="toctree-l1"><a class="reference internal" href="gsa-example.html">Gene set analysis example</a></li>
<li class="toctree-l1"><a class="reference internal" href="pathway-example.html">Pathway example</a></li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References</a></li>
</ul>

          <h3 style="margin-top: 1.5em;">Search</h3>
          <form class="search" action="search.html" method="get">
            <input type="text" name="q" />
            <input type="submit" value="Go" />
            <input type="hidden" name="check_keywords" value="yes" />
            <input type="hidden" name="area" value="default" />
          </form>
          <p class="searchtip" style="font-size: 90%">
            Enter search terms or a module, class or function name.
          </p>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

    <div class="footer-wrapper">
      <div class="footer">
        <div class="left">
          <a href="install.html" title="Installation"
             >previous</a> |
          <a href="database-cookbook.html" title="Common database tasks"
             >next</a> |
          <a href="genindex.html" title="General Index"
             >index</a>
            <br/>
            <a href="_sources/database.txt"
               rel="nofollow">Show Source</a>
        </div>

        <div class="right">
          
    <div class="footer">
        &copy; Copyright 2015, A. Richards.
      Last updated on Oct 13, 2015.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.2.
    </div>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

  </body>
</html>