<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Gene set analysis &mdash; hts-integrate - 0.4.1</title>
    
    <link rel="stylesheet" href="_static/agogo.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="hts-integrate - 0.4.1" href="index.html" />
    <link rel="next" title="Works cited" href="references.html" />
    <link rel="prev" title="Differential expression analysis" href="deseq-example.html" /> 
  </head>
  <body>
    <div class="header-wrapper">
      <div class="header">
        <div class="headertitle"><a
          href="index.html">hts-integrate - 0.4.1</a></div>
        <div class="rel">
          <a href="deseq-example.html" title="Differential expression analysis"
             accesskey="P">previous</a> |
          <a href="references.html" title="Works cited"
             accesskey="N">next</a> |
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a>
        </div>
       </div>
    </div>

    <div class="content-wrapper">
      <div class="content">
        <div class="document">
            
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="gene-set-analysis">
<h1>Gene set analysis<a class="headerlink" href="#gene-set-analysis" title="Permalink to this headline">¶</a></h1>
<p>The procedure in this tutorial example to carry out gene set analysis using <tt class="docutils literal"><span class="pre">htsint</span></tt> is composed of three main steps.</p>
<blockquote>
<div><ol class="arabic simple">
<li>Gene set generation</li>
<li>Gene set testing</li>
<li>Gene set visualization</li>
</ol>
</div></blockquote>
<p>This example takes approximately 30 minutes to complete on a modern desktop computer.  The code discussed in this document is available as a script for convenience.</p>
<blockquote>
<div><ul class="simple">
<li><a class="reference download internal" href="_downloads/gsa-example.py"><tt class="xref download docutils literal"><span class="pre">gsa-example.py</span></tt></a></li>
</ul>
</div></blockquote>
<div class="section" id="gene-set-generation">
<h2>Gene set generation<a class="headerlink" href="#gene-set-generation" title="Permalink to this headline">¶</a></h2>
<p>The basic process involves integrating Gene Ontology <a class="reference internal" href="references.html#ashburner00" id="id1">[Ashburner00]</a> information from one or more taxa to infer functional distances between genes.  These distances are then used to cluster the genes in a specified list.  These genes are any groups of genes that you might want to cluster.  For example, you may use the genes in a large pathway or all genes from an RNA-Seq experiment, as is the case in this example.</p>
<ol class="arabic">
<li><p class="first">Run <a class="reference internal" href="blast.html"><em>BLAST and create a summarized blast map</em></a>.  To save time in this tutorial we provide an example summary file below.</p>
<ul class="simple">
<li><a class="reference download internal" href="_downloads/blast-parsed-summary.csv"><tt class="xref download docutils literal"><span class="pre">blast-parsed-summary.csv</span></tt></a></li>
</ul>
<p>Load the file.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">htsint.blast</span> <span class="kn">import</span> <span class="n">BlastMapper</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bm</span> <span class="o">=</span> <span class="n">BlastMapper</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bmap</span> <span class="o">=</span> <span class="n">bm</span><span class="o">.</span><span class="n">load_summary</span><span class="p">(</span><span class="s">&#39;blast-parsed-summary.csv&#39;</span><span class="p">,</span><span class="n">best</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
</pre></div>
</div>
</li>
</ol>
<div class="section" id="create-a-term-graph">
<h3>Create a term graph<a class="headerlink" href="#create-a-term-graph" title="Permalink to this headline">¶</a></h3>
<p>Because genes and their ontology terms will be loaded multiple times fetch the annotations only once then save the dictionaries.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">htsint</span> <span class="kn">import</span> <span class="n">GeneOntology</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">go</span> <span class="o">=</span> <span class="n">GeneOntology</span><span class="p">([</span><span class="s">&quot;8364&quot;</span><span class="p">,</span><span class="s">&quot;8355&quot;</span><span class="p">],</span><span class="n">useIea</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span><span class="n">aspect</span><span class="o">=</span><span class="s">&#39;molecular_function&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">termsPath</span> <span class="o">=</span> <span class="s">&quot;go-terms.pickle&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">graphPath</span> <span class="o">=</span> <span class="s">&quot;go-graph.pickle&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">go</span><span class="o">.</span><span class="n">create_dicts</span><span class="p">(</span><span class="n">termsPath</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gene2go</span><span class="p">,</span><span class="n">go2gene</span> <span class="o">=</span> <span class="n">go</span><span class="o">.</span><span class="n">load_dicts</span><span class="p">(</span><span class="n">termsPath</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">G</span> <span class="o">=</span> <span class="n">go</span><span class="o">.</span><span class="n">create_gograph</span><span class="p">(</span><span class="n">termsPath</span><span class="o">=</span><span class="n">termsPath</span><span class="p">,</span><span class="n">graphPath</span><span class="o">=</span><span class="n">graphPath</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="s">&quot;</span><span class="si">%s</span><span class="s"> genes have at least one annotation&quot;</span><span class="o">%</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gene2go</span><span class="o">.</span><span class="n">keys</span><span class="p">())))</span>
<span class="go">Term graph for with 15719 nodes successfully created.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="s">&quot;Term graph for with </span><span class="si">%s</span><span class="s"> nodes successfully created.&quot;</span><span class="o">%</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">G</span><span class="o">.</span><span class="n">nodes</span><span class="p">())))</span>
<span class="go">1291 genes have at least one annotation</span>
</pre></div>
</div>
<p>The valid GO aspects are: &#8216;biological_process&#8217;, &#8216;molecular_function&#8217; and &#8216;cellular_component&#8217;.</p>
</div>
<div class="section" id="calculate-term-distances">
<h3>Calculate term distances<a class="headerlink" href="#calculate-term-distances" title="Permalink to this headline">¶</a></h3>
<p>Ideally, this step is carried out in a cluster environment and if you are using <a class="reference external" href="http://gridscheduler.sourceforge.net">Grid Engine</a> then there are built-in convenience methods.  Whether you are in a high performance environment or on a single machine the initialization is the same.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">htsint</span> <span class="kn">import</span> <span class="n">TermDistances</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">td</span> <span class="o">=</span> <span class="n">TermDistances</span><span class="p">(</span><span class="n">termsPath</span><span class="p">,</span><span class="n">graphPath</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">termDistancePath</span> <span class="o">=</span> <span class="s">&quot;term-distances.csv&quot;</span>
</pre></div>
</div>
<div class="section" id="using-grid-engine">
<h4>Using Grid Engine:<a class="headerlink" href="#using-grid-engine" title="Permalink to this headline">¶</a></h4>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">cpus</span> <span class="o">=</span> <span class="mi">60</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">td</span><span class="o">.</span><span class="n">create_scripts</span><span class="p">(</span><span class="s">&#39;youremail@somewhere.edu&#39;</span><span class="p">,</span><span class="n">cpus</span><span class="o">=</span><span class="n">cpus</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">td</span><span class="o">.</span><span class="n">submit</span><span class="p">()</span>
</pre></div>
</div>
<p>Before you submit you can check in the <tt class="docutils literal"><span class="pre">htsint-tmp</span></tt> directory that was created in the current working directory to ensure the Bash scripts work for your computing environment.  The results are then assembled into a single file.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">htsint</span> <span class="kn">import</span> <span class="n">AssembleDistances</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ad</span> <span class="o">=</span> <span class="n">AssembleDistances</span><span class="p">(</span><span class="n">termsPath</span><span class="p">,</span><span class="n">resultsPath</span><span class="o">=</span><span class="n">termDistancePath</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ad</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">cpus</span><span class="o">=</span><span class="n">cpus</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="using-single-machine">
<h4>Using single machine<a class="headerlink" href="#using-single-machine" title="Permalink to this headline">¶</a></h4>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">td</span><span class="o">.</span><span class="n">run_with_multiprocessing</span><span class="p">(</span><span class="n">termDistancePath</span><span class="p">,</span><span class="n">cpus</span><span class="o">=</span><span class="mi">16</span><span class="p">)</span>
</pre></div>
</div>
<p>This is the most computationally expensive step in the pipeline so for lists with more than a few thousand genes this calculation becomes difficult outside of a cluster environment.  Using 16 cores on a single machine the previous command finished in 00:27:32 (hh:mm:ss).</p>
</div>
</div>
<div class="section" id="calculate-gene-distances">
<h3>Calculate gene distances<a class="headerlink" href="#calculate-gene-distances" title="Permalink to this headline">¶</a></h3>
<p>With the term-term distances stored in the distance file we can map the gene-gene distances.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">htsint</span> <span class="kn">import</span> <span class="n">GeneDistances</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">geneDistancePath</span> <span class="o">=</span> <span class="s">&quot;gene-distances.csv&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gd</span> <span class="o">=</span> <span class="n">GeneDistances</span><span class="p">(</span><span class="n">termsPath</span><span class="p">,</span><span class="n">graphPath</span><span class="p">,</span><span class="n">termDistancePath</span><span class="p">,</span><span class="n">outFile</span><span class="o">=</span><span class="n">geneDistancePath</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gd</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="section" id="spectral-clustering">
<h3>Spectral Clustering<a class="headerlink" href="#spectral-clustering" title="Permalink to this headline">¶</a></h3>
<p>With the gene-gene distances a number of unsupervised clustering algorithms can be used here.  [Completed on next update]</p>
</div>
<div class="section" id="save-gene-sets">
<h3>Save gene sets<a class="headerlink" href="#save-gene-sets" title="Permalink to this headline">¶</a></h3>
</div>
</div>
</div>


          </div>
        </div>
      </div>
        </div>
        <div class="sidebar">
          <h3>Table Of Contents</h3>
          <ul>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="database.html">Database</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="database-cookbook.html">Database cookbook</a></li>
<li class="toctree-l1"><a class="reference internal" href="blast.html">Using BLAST</a></li>
<li class="toctree-l1"><a class="reference internal" href="annotation-fetching.html">Gene Ontology annotation fetching</a></li>
<li class="toctree-l1"><a class="reference internal" href="deseq-example.html">Gene level differential expression</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="">Gene set analysis example</a></li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References</a></li>
</ul>

          <h3 style="margin-top: 1.5em;">Search</h3>
          <form class="search" action="search.html" method="get">
            <input type="text" name="q" />
            <input type="submit" value="Go" />
            <input type="hidden" name="check_keywords" value="yes" />
            <input type="hidden" name="area" value="default" />
          </form>
          <p class="searchtip" style="font-size: 90%">
            Enter search terms or a module, class or function name.
          </p>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

    <div class="footer-wrapper">
      <div class="footer">
        <div class="left">
          <a href="deseq-example.html" title="Differential expression analysis"
             >previous</a> |
          <a href="references.html" title="Works cited"
             >next</a> |
          <a href="genindex.html" title="General Index"
             >index</a>
            <br/>
            <a href="_sources/gsa-example.txt"
               rel="nofollow">Show Source</a>
        </div>

        <div class="right">
          
    <div class="footer">
        &copy; Copyright 2015, A. Richards.
      Last updated on Mar 26, 2015.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.2.
    </div>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

  </body>
</html>