<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Gene set analysis &mdash; htsint - 0.5.1</title>
    
    <link rel="stylesheet" href="_static/agogo.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="htsint - 0.5.1" href="index.html" />
    <link rel="next" title="Pathway example" href="pathway-example.html" />
    <link rel="prev" title="Differential expression analysis" href="deseq-example.html" /> 
  </head>
  <body>
    <div class="header-wrapper">
      <div class="header">
        <div class="headertitle"><a
          href="index.html">htsint - 0.5.1</a></div>
        <div class="rel">
          <a href="deseq-example.html" title="Differential expression analysis"
             accesskey="P">previous</a> |
          <a href="pathway-example.html" title="Pathway example"
             accesskey="N">next</a> |
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a>
        </div>
       </div>
    </div>

    <div class="content-wrapper">
      <div class="content">
        <div class="document">
            
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="gene-set-analysis">
<h1>Gene set analysis<a class="headerlink" href="#gene-set-analysis" title="Permalink to this headline">¶</a></h1>
<p>The procedure in this tutorial example to carry out gene set analysis using <tt class="docutils literal"><span class="pre">htsint</span></tt> is composed of three main steps.</p>
<blockquote>
<div><ol class="arabic simple">
<li>Gene set generation</li>
<li>Gene set testing</li>
<li>Gene set visualization</li>
</ol>
</div></blockquote>
<p>This example takes anywhere from 30 minutes to several hours on a modern desktop computer.  To run a shorter example use the <tt class="docutils literal"><span class="pre">molecular_function</span></tt> domain of the Gene Ontology.  The code used for gene set generation in this document is available as a script for convenience.</p>
<blockquote>
<div><ul class="simple">
<li><a class="reference download internal" href="_downloads/gsa-example.py"><tt class="xref download docutils literal"><span class="pre">gsa-example.py</span></tt></a></li>
</ul>
</div></blockquote>
<div class="section" id="gene-set-generation">
<h2>(1) Gene set generation<a class="headerlink" href="#gene-set-generation" title="Permalink to this headline">¶</a></h2>
<p>The basic process involves integrating Gene Ontology <a class="reference internal" href="references.html#ashburner00" id="id1">[Ashburner00]</a> information from one or more taxa to infer functional distances between genes.  These distances are then used to cluster the genes in a specified list.  These genes are any groups of genes that you might want to cluster.  For example, you may use the genes in a large pathway or all genes from an RNA-Seq experiment, as is the case in this example.</p>
<div class="section" id="create-a-term-graph">
<h3>Create a term graph<a class="headerlink" href="#create-a-term-graph" title="Permalink to this headline">¶</a></h3>
<p>First make the necessary imports.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">time</span><span class="o">,</span><span class="nn">os</span><span class="o">,</span><span class="nn">re</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">htsint</span> <span class="kn">import</span> <span class="n">GeneOntology</span><span class="p">,</span><span class="n">TermDistances</span><span class="p">,</span><span class="n">GeneDistances</span><span class="p">,</span><span class="n">GeneSetCollection</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">htsint.stats</span> <span class="kn">import</span> <span class="n">SpectralCluster</span><span class="p">,</span> <span class="n">SpectralClusterParamSearch</span><span class="p">,</span> <span class="n">SpectralClusterResults</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">htsint.blast</span> <span class="kn">import</span> <span class="n">BlastMapper</span>
</pre></div>
</div>
<p>Then specify an output directory along with a GO aspect with an abbreviation.  The valid GO aspects are: &#8216;biological_process&#8217;, &#8216;molecular_function&#8217; and &#8216;cellular_component&#8217;.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">homeDir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s">&quot;.&quot;</span><span class="p">,</span><span class="s">&quot;demo&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">homeDir</span><span class="p">):</span>
<span class="gp">&gt;&gt;&gt; </span> <span class="n">os</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">homeDir</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">aspect</span> <span class="o">=</span> <span class="s">&#39;biological_process&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_aspect</span> <span class="o">=</span> <span class="s">&#39;bp&#39;</span>
</pre></div>
</div>
<p>Because genes and their ontology terms will be loaded multiple times fetch the annotations only once then save the dictionaries.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">go</span> <span class="o">=</span> <span class="n">GeneOntology</span><span class="p">([</span><span class="s">&quot;8364&quot;</span><span class="p">,</span><span class="s">&quot;8355&quot;</span><span class="p">],</span><span class="n">useIea</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span><span class="n">aspect</span><span class="o">=</span><span class="n">aspect</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">termsPath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">homeDir</span><span class="p">,</span><span class="s">&quot;go-terms.pickle&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">graphPath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">homeDir</span><span class="p">,</span><span class="s">&quot;go-graph.pickle&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">go</span><span class="o">.</span><span class="n">create_dicts</span><span class="p">(</span><span class="n">termsPath</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gene2go</span><span class="p">,</span><span class="n">go2gene</span> <span class="o">=</span> <span class="n">go</span><span class="o">.</span><span class="n">load_dicts</span><span class="p">(</span><span class="n">termsPath</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="s">&quot;</span><span class="si">%s</span><span class="s"> genes have at least one annotation&quot;</span><span class="o">%</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gene2go</span><span class="o">.</span><span class="n">keys</span><span class="p">())))</span>
<span class="go">1597 genes have at least one annotation</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="s">&quot;Term graph for with </span><span class="si">%s</span><span class="s"> nodes successfully created.&quot;</span><span class="o">%</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">G</span><span class="o">.</span><span class="n">nodes</span><span class="p">())))</span>
<span class="go">total distances to evaluate: 1219141.0</span>
</pre></div>
</div>
<p>Create the graph of term-term relationships.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">G</span> <span class="o">=</span> <span class="n">go</span><span class="o">.</span><span class="n">create_gograph</span><span class="p">(</span><span class="n">termsPath</span><span class="o">=</span><span class="n">termsPath</span><span class="p">,</span><span class="n">graphPath</span><span class="o">=</span><span class="n">graphPath</span><span class="p">)</span>
<span class="go">Term graph for with 15719 nodes successfully created.</span>
</pre></div>
</div>
</div>
<div class="section" id="calculate-term-distances">
<h3>Calculate term distances<a class="headerlink" href="#calculate-term-distances" title="Permalink to this headline">¶</a></h3>
<p>Ideally, this step is carried out in a cluster environment and if you are using <a class="reference external" href="http://gridscheduler.sourceforge.net">Grid Engine</a> then there are built-in convenience methods.  Whether you are in a high performance environment or on a single machine the initialization is the same.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">termDistancePath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">homeDir</span><span class="p">,</span><span class="s">&quot;term-distances.npy&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">td</span> <span class="o">=</span> <span class="n">TermDistances</span><span class="p">(</span><span class="n">termsPath</span><span class="p">,</span><span class="n">graphPath</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="s">&quot;total distances to evaluate: </span><span class="si">%s</span><span class="s">&quot;</span><span class="o">%</span><span class="n">td</span><span class="o">.</span><span class="n">totalDistances</span><span class="p">)</span>
<span class="go">total distances to evaluate: 1219141.0</span>
</pre></div>
</div>
<div class="section" id="using-grid-engine">
<h4>Using Grid Engine:<a class="headerlink" href="#using-grid-engine" title="Permalink to this headline">¶</a></h4>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">cpus</span> <span class="o">=</span> <span class="mi">60</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">td</span><span class="o">.</span><span class="n">create_scripts</span><span class="p">(</span><span class="s">&#39;youremail@somewhere.edu&#39;</span><span class="p">,</span><span class="n">cpus</span><span class="o">=</span><span class="n">cpus</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">td</span><span class="o">.</span><span class="n">submit</span><span class="p">()</span>
</pre></div>
</div>
<p>Before you submit you can check in the <tt class="docutils literal"><span class="pre">htsint-tmp</span></tt> directory that was created in the current working directory to ensure the Bash scripts work for your computing environment.  The results are then assembled into a single file.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">htsint</span> <span class="kn">import</span> <span class="n">AssembleDistances</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ad</span> <span class="o">=</span> <span class="n">AssembleDistances</span><span class="p">(</span><span class="n">termsPath</span><span class="p">,</span><span class="n">resultsPath</span><span class="o">=</span><span class="n">termDistancePath</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ad</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">cpus</span><span class="o">=</span><span class="n">cpus</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="using-single-machine">
<h4>Using single machine<a class="headerlink" href="#using-single-machine" title="Permalink to this headline">¶</a></h4>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">td</span><span class="o">.</span><span class="n">run_with_multiprocessing</span><span class="p">(</span><span class="n">termDistancePath</span><span class="p">,</span><span class="n">cpus</span><span class="o">=</span><span class="mi">16</span><span class="p">)</span>
</pre></div>
</div>
<p>This is the most computationally expensive step in the pipeline so for lists with more than a few thousand genes this calculation becomes difficult outside of a cluster environment.  Using 8 cores on a single machine the previous command finished in XX:XX:XX (hh:mm:ss).</p>
</div>
</div>
<div class="section" id="calculate-gene-distances">
<h3>Calculate gene distances<a class="headerlink" href="#calculate-gene-distances" title="Permalink to this headline">¶</a></h3>
<p>With the term-term distances stored in the distance file we can map the gene-gene distances.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">geneDistancePath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">homeDir</span><span class="p">,</span><span class="s">&quot;gene-distances.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gd</span> <span class="o">=</span> <span class="n">GeneDistances</span><span class="p">(</span><span class="n">termsPath</span><span class="p">,</span><span class="n">termDistancePath</span><span class="p">,</span><span class="n">outFile</span><span class="o">=</span><span class="n">geneDistancePath</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gd</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="section" id="spectral-clustering">
<h3>Spectral Clustering<a class="headerlink" href="#spectral-clustering" title="Permalink to this headline">¶</a></h3>
<p>With the gene-gene distances a number of unsupervised clustering algorithms can be used here.  Because spectral clustering is appropriate for networks we have implemented two algorithms as part of <tt class="docutils literal"><span class="pre">htsint</span></tt>.  There is a bandwidth parameter <img class="math" src="_images/math/2298cf1485084afe72757a9c8483af49a138d81f.png" alt="\sigma"/> and a parameter for the number of clusters <cite>k</cite> that need to be given.</p>
<div class="section" id="parameter-estimation-optional">
<h4>Parameter estimation [optional]<a class="headerlink" href="#parameter-estimation-optional" title="Permalink to this headline">¶</a></h4>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">silvalFile</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="s">&quot;\.csv&quot;</span><span class="p">,</span><span class="s">&quot;-scparams-sv.csv&quot;</span><span class="p">,</span><span class="n">geneDistancePath</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">clustersFile</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="s">&quot;\.csv&quot;</span><span class="p">,</span><span class="s">&quot;-scparams-cl.csv&quot;</span><span class="p">,</span><span class="n">geneDistancePath</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">scps</span> <span class="o">=</span> <span class="n">SpectralClusterParamSearch</span><span class="p">(</span><span class="n">geneDistancePath</span><span class="p">,</span><span class="n">dtype</span><span class="o">=</span><span class="s">&#39;distance&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">scps</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">chunks</span><span class="o">=</span><span class="mi">15</span><span class="p">)</span>
</pre></div>
</div>
<p>Plot the parameter search</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">psFigureFile</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">homeDir</span><span class="p">,</span><span class="s">&quot;param-scan-</span><span class="si">%s</span><span class="s">.png&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">_aspect</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">scr</span> <span class="o">=</span> <span class="n">SpectralClusterResults</span><span class="p">(</span><span class="n">silvalFile</span><span class="p">,</span><span class="n">clustersFile</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">scr</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">figName</span><span class="o">=</span><span class="n">psFigureFile</span><span class="p">)</span>
</pre></div>
</div>
<div class="align-center figure">
<a class="reference internal image-reference" href="_images/param-scan-bp.png"><img alt="parameter scan" src="_images/param-scan-bp.png" style="width: 700.0px; height: 600.0px;" /></a>
</div>
<p>Ideally, we are looking for values of <img class="math" src="_images/math/2298cf1485084afe72757a9c8483af49a138d81f.png" alt="\sigma"/> and <cite>k</cite> that maximize our silhouette value, while at the same time maximize the number of clusters that fall into a reasonable size range.  The size range can be set with the <tt class="docutils literal"><span class="pre">threshMin</span></tt> and <tt class="docutils literal"><span class="pre">threshMax</span></tt> arguments.  It helps result interpretation if the specified range can be reasonably investigated through visualization.  The top panel shows the average silhouette value for the clustering results over a grid of possible parameter values. For the same grid the bottom panel illustrates the percentage of total genes that fall into clusters of the desired size.  There is usually a trade-off between high silhouette values and the reasonably sized clusters.  The top three optimal values are marked on the plots.  For this example the parameters are maximized at <img class="math" src="_images/math/3b2f575eb4e7de2f7e39fc2cc2986fac142e0475.png" alt="k=123"/> and <img class="math" src="_images/math/fe818f5845875abf4a1b8bc1ab59e6325ffb9ef6.png" alt="\sigma=0.08"/>.  It is worth noting that strongly associated clusters tend to remain mostly intact over a wide range of parameter values.  In the script version of this example this section the parameter estimation is commented out to minimize compute time.</p>
</div>
<div class="section" id="run-spectral-clustering">
<h4>Run spectral clustering<a class="headerlink" href="#run-spectral-clustering" title="Permalink to this headline">¶</a></h4>
<p>There are two implementations of spectral clustering available through the SpectralCluster class.  If the argument <tt class="docutils literal"><span class="pre">sk</span></tt> is <tt class="docutils literal"><span class="pre">None</span></tt> then the original algorithm proposed by Andrew Ng <em>et al</em>. is used <a class="reference internal" href="references.html#ng01" id="id2">[Ng01]</a>.  Alternatively, a self-tuning version of this algorithm was proposed by Zelnik-Manor and Perona that uses a different <img class="math" src="_images/math/2298cf1485084afe72757a9c8483af49a138d81f.png" alt="\sigma"/> around each neighborhood.  The neighborhood size is controlled by the parameter <tt class="docutils literal"><span class="pre">sk</span></tt> as discussed in the manuscript <a class="reference internal" href="references.html#zelnik-manor04" id="id3">[Zelnik-Manor04]</a>.  For smaller networks the self-tuning method gives reasonable results, however for larger networks the grid parameter search seems to provide more biologically intuitive clusters.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">htsint.stats</span> <span class="kn">import</span> <span class="n">SpectralCluster</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">k</span> <span class="o">=</span> <span class="mi">123</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sigma</span> <span class="o">=</span> <span class="mf">0.08</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span> <span class="o">=</span> <span class="n">SpectralCluster</span><span class="p">(</span><span class="n">geneDistancePath</span><span class="p">,</span><span class="n">dtype</span><span class="o">=</span><span class="s">&#39;distance&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">k</span><span class="p">,</span><span class="n">sk</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span><span class="n">sigma</span><span class="o">=</span><span class="n">sigma</span><span class="p">,</span><span class="n">verbose</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">labelsPath</span><span class="o">=</span><span class="n">labelsPath</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="map-genes-to-transcripts-and-save-gene-sets">
<h3>Map genes to transcripts and save gene sets<a class="headerlink" href="#map-genes-to-transcripts-and-save-gene-sets" title="Permalink to this headline">¶</a></h3>
<p>Run <a class="reference internal" href="blast.html"><em>BLAST and create a summarized blast map</em></a>.  To save time in this tutorial we provide an example summary file below.</p>
<blockquote>
<div><ul class="simple">
<li><a class="reference download internal" href="_downloads/blast-parsed-summary.csv"><tt class="xref download docutils literal"><span class="pre">blast-parsed-summary.csv</span></tt></a></li>
</ul>
<p>Load the file.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">htsint.blast</span> <span class="kn">import</span> <span class="n">BlastMapper</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bm</span> <span class="o">=</span> <span class="n">BlastMapper</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bmap</span> <span class="o">=</span> <span class="n">bm</span><span class="o">.</span><span class="n">load_summary</span><span class="p">(</span><span class="s">&#39;blast-parsed-summary.csv&#39;</span><span class="p">,</span><span class="n">best</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
</pre></div>
</div>
</div></blockquote>
<p>Then we can constrain the gene set size in terms of the number of transcripts by setting the <tt class="docutils literal"><span class="pre">transcriptMin</span></tt> and <tt class="docutils literal"><span class="pre">transcriptMax</span></tt> arguments.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">htsint</span> <span class="kn">import</span> <span class="n">GeneSetCollection</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">transcriptMin</span><span class="p">,</span><span class="n">transcriptMax</span> <span class="o">=</span> <span class="mi">9</span><span class="p">,</span><span class="mi">1000</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gsFile</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">homeDir</span><span class="p">,</span><span class="s">&quot;</span><span class="si">%s</span><span class="s">.gmt&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">_aspect</span><span class="p">))</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">gsc</span> <span class="o">=</span> <span class="n">GeneSetCollection</span><span class="p">(</span><span class="n">labelsPath</span><span class="p">,</span><span class="n">gene2go</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gsc</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">blastMap</span><span class="o">=</span><span class="n">bmap</span><span class="p">,</span><span class="n">transcriptMin</span><span class="o">=</span><span class="n">transcriptMin</span><span class="p">,</span><span class="n">transcriptMax</span><span class="o">=</span><span class="n">transcriptMax</span><span class="p">,</span><span class="n">outFile</span><span class="o">=</span><span class="n">gsFile</span><span class="p">)</span>
</pre></div>
</div>
<p>The class <tt class="docutils literal"><span class="pre">GeneSetCollection</span></tt> will create two files and the <tt class="docutils literal"><span class="pre">write</span></tt> method can be used to create these files for different assemblies assuming the BLAST mapper is appropriate.  The file name specified by gsFile will specify where to write <a class="reference external" href="http://www.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#GMT:_Gene_Matrix_Transposed_file_format_.28.2A.gmt.29">GMT formatted</a> file.  For convenience and visualization another file (in csv format) with the same  name base will be created to store the gene to transcript mappings.  In this case <tt class="docutils literal"><span class="pre">./demo/bp.gmt</span></tt> and <tt class="docutils literal"><span class="pre">./demo/bp.csv</span></tt> will be created.</p>
</div>
</div>
<div class="section" id="gene-set-testing-gsa">
<h2>(2) Gene set testing (GSA)<a class="headerlink" href="#gene-set-testing-gsa" title="Permalink to this headline">¶</a></h2>
<p>There are a number of packages available to run Gene set analysis.  We use in this example the <a class="reference external" href="http://cran.r-project.org/web/packages/GSA/GSA.pdf">GSA R package</a>.  Download and run the following script.</p>
<blockquote>
<div><a class="reference download internal" href="_downloads/run-gsa.R"><tt class="xref download docutils literal"><span class="pre">run-gsa.R</span></tt></a></div></blockquote>
<p>Before you can run the R script a count file (<tt class="docutils literal"><span class="pre">deseq-samples.csv</span></tt>) containing transformed counts needs to be in the current working directory.  If you followed the <a class="reference internal" href="deseq-example.html"><em>gene expression tutorial</em></a> then the file should be present.  Otherwise, download the <a class="reference download internal" href="_downloads/raw-counts.csv"><tt class="xref download docutils literal"><span class="pre">raw-counts.csv</span></tt></a> and  <a class="reference download internal" href="_downloads/run-deseq.R"><tt class="xref download docutils literal"><span class="pre">run-deseq.R</span></tt></a> scripts and run the following first.</p>
<p>To run the example you use:</p>
<blockquote>
<div><div class="highlight-bash"><div class="highlight"><pre>~<span class="nv">$ </span>Rscript run-deseq.R raw-counts.csv
</pre></div>
</div>
</div></blockquote>
<p>Once you have run DESeq the transformed counts are available and GSA can be run.</p>
<blockquote>
<div><div class="highlight-bash"><div class="highlight"><pre>~<span class="nv">$ </span>Rscript run-gsa.R bp
</pre></div>
</div>
</div></blockquote>
<p>Inside of the demo directory a file names <tt class="docutils literal"><span class="pre">geneset-results.csv</span></tt> will be created identifing the gene sets with FDR values less than 0.5.  Edit the script and refer to the documentation to modify the file to fit you experimental setting.</p>
</div>
<div class="section" id="visualizing-gene-sets">
<h2>(3) Visualizing gene sets<a class="headerlink" href="#visualizing-gene-sets" title="Permalink to this headline">¶</a></h2>
<p>First, we need to specify a few paths to files that were created duing the gene set generation process.  Then using the object <tt class="docutils literal"><span class="pre">GeneSet</span></tt>, a network visualization that makes use of <a class="reference external" href="http://networkx.github.io/">NetworkX</a> can be made for any arbitrary gene set.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="kn">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">htsint</span> <span class="kn">import</span> <span class="n">GeneSet</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">distMat</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s">&quot;.&quot;</span><span class="p">,</span><span class="s">&quot;demo&quot;</span><span class="p">,</span><span class="s">&quot;term-distances.npy&quot;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">genesetFile</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s">&quot;.&quot;</span><span class="p">,</span><span class="s">&quot;demo&quot;</span><span class="p">,</span><span class="s">&quot;bp.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">termsPath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s">&quot;.&quot;</span><span class="p">,</span><span class="s">&quot;demo&quot;</span><span class="p">,</span><span class="s">&quot;go-terms.pickle&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gsets</span> <span class="o">=</span> <span class="n">GeneSet</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gsets</span><span class="o">.</span><span class="n">load_geneset</span><span class="p">(</span><span class="n">genesetFile</span><span class="p">,</span><span class="n">termsPath</span><span class="p">,</span><span class="n">distMat</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">genesetId</span> <span class="o">=</span> <span class="s">&#39;gs-60&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gsets</span><span class="o">.</span><span class="n">draw_figure</span><span class="p">(</span><span class="n">genesetId</span><span class="p">,</span><span class="n">layout</span><span class="o">=</span><span class="s">&#39;spring&#39;</span><span class="p">,</span><span class="n">name</span><span class="o">=</span><span class="s">&#39;</span><span class="si">%s</span><span class="s">.png&#39;</span><span class="o">%</span><span class="p">(</span><span class="n">genesetId</span><span class="p">),</span><span class="n">percentile</span><span class="o">=</span><span class="mi">25</span><span class="p">)</span>
</pre></div>
</div>
<div class="align-center figure">
<a class="reference internal image-reference" href="_images/gs-60.png"><img alt="example gene set" src="_images/gs-60.png" style="width: 612.0px; height: 372.0px;" /></a>
</div>
</div>
</div>


          </div>
        </div>
      </div>
        </div>
        <div class="sidebar">
          <h3>Table Of Contents</h3>
          <ul>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="database.html">Database</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="database-cookbook.html">Database cookbook</a></li>
<li class="toctree-l1"><a class="reference internal" href="blast.html">Using BLAST</a></li>
<li class="toctree-l1"><a class="reference internal" href="annotation-fetching.html">Gene Ontology annotation fetching</a></li>
<li class="toctree-l1"><a class="reference internal" href="deseq-example.html">Gene level differential expression</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="">Gene set analysis example</a></li>
<li class="toctree-l1"><a class="reference internal" href="pathway-example.html">Pathway example</a></li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References</a></li>
</ul>

          <h3 style="margin-top: 1.5em;">Search</h3>
          <form class="search" action="search.html" method="get">
            <input type="text" name="q" />
            <input type="submit" value="Go" />
            <input type="hidden" name="check_keywords" value="yes" />
            <input type="hidden" name="area" value="default" />
          </form>
          <p class="searchtip" style="font-size: 90%">
            Enter search terms or a module, class or function name.
          </p>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

    <div class="footer-wrapper">
      <div class="footer">
        <div class="left">
          <a href="deseq-example.html" title="Differential expression analysis"
             >previous</a> |
          <a href="pathway-example.html" title="Pathway example"
             >next</a> |
          <a href="genindex.html" title="General Index"
             >index</a>
            <br/>
            <a href="_sources/gsa-example.txt"
               rel="nofollow">Show Source</a>
        </div>

        <div class="right">
          
    <div class="footer">
        &copy; Copyright 2015, A. Richards.
      Last updated on Aug 21, 2015.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.2.
    </div>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

  </body>
</html>