<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Pathway example &mdash; htsint - 0.5.1</title>
    
    <link rel="stylesheet" href="_static/agogo.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="htsint - 0.5.1" href="index.html" />
    <link rel="next" title="Works cited" href="references.html" />
    <link rel="prev" title="Gene set analysis" href="gsa-example.html" /> 
  </head>
  <body>
    <div class="header-wrapper">
      <div class="header">
        <div class="headertitle"><a
          href="index.html">htsint - 0.5.1</a></div>
        <div class="rel">
          <a href="gsa-example.html" title="Gene set analysis"
             accesskey="P">previous</a> |
          <a href="references.html" title="Works cited"
             accesskey="N">next</a> |
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a>
        </div>
       </div>
    </div>

    <div class="content-wrapper">
      <div class="content">
        <div class="document">
            
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="pathway-example">
<h1>Pathway example<a class="headerlink" href="#pathway-example" title="Permalink to this headline">Â¶</a></h1>
<p>Lets start with a pathway downloaded from the KEGG database.</p>
<blockquote>
<div><ul class="simple">
<li><a class="reference download internal" href="_downloads/hsa00860.txt"><tt class="xref download docutils literal"><span class="pre">hsa00860.txt</span></tt></a></li>
</ul>
</div></blockquote>
<p>The Python code in this example is available as a script.</p>
<blockquote>
<div><ul class="simple">
<li><a class="reference download internal" href="_downloads/pathway-example.py"><tt class="xref download docutils literal"><span class="pre">pathway-example.py</span></tt></a></li>
</ul>
</div></blockquote>
<p>First lets define a parsing function for the KEGG formatted pathways file.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="k">def</span> <span class="nf">get_genes</span><span class="p">(</span><span class="n">pathway</span><span class="p">):</span>
    <span class="n">fid</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s">&quot;.&quot;</span><span class="p">,</span><span class="n">pathway</span><span class="o">+</span><span class="s">&quot;.txt&quot;</span><span class="p">),</span><span class="s">&#39;r&#39;</span><span class="p">)</span>
    <span class="n">isGene</span> <span class="o">=</span> <span class="bp">False</span>
    <span class="n">gene2symbol</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">linja</span> <span class="ow">in</span> <span class="n">fid</span><span class="p">:</span>
        <span class="n">linja</span> <span class="o">=</span> <span class="n">linja</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s">&quot;^GENE&quot;</span><span class="p">,</span><span class="n">linja</span><span class="p">):</span>
            <span class="n">isGene</span> <span class="o">=</span> <span class="bp">True</span>
        <span class="k">elif</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s">&quot;^[A-Z]+&quot;</span><span class="p">,</span><span class="n">linja</span><span class="p">):</span>
            <span class="n">isGene</span> <span class="o">=</span> <span class="bp">False</span>
        <span class="k">if</span> <span class="n">isGene</span><span class="p">:</span>
            <span class="n">geneIds</span> <span class="o">=</span> <span class="n">linja</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot;;&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">_geneIds</span> <span class="o">=</span> <span class="p">[</span><span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="s">&quot;\s+&quot;</span><span class="p">,</span><span class="s">&quot;&quot;</span><span class="p">,</span><span class="n">gi</span><span class="p">)</span> <span class="k">for</span> <span class="n">gi</span> <span class="ow">in</span> <span class="n">geneIds</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot;  &quot;</span><span class="p">)]</span>
            <span class="n">geneIds</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">gid</span> <span class="ow">in</span> <span class="n">_geneIds</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">gid</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="n">gid</span> <span class="o">!=</span> <span class="s">&#39;GENE&#39;</span><span class="p">:</span>
                    <span class="n">geneIds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gid</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">geneIds</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">2</span><span class="p">:</span>
                <span class="k">print</span><span class="p">(</span><span class="s">&quot;...</span><span class="si">%s</span><span class="s">&quot;</span><span class="o">%</span><span class="nb">str</span><span class="p">(</span><span class="n">geneIds</span><span class="p">))</span>
                <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s">&quot;Could not parse gene names&quot;</span><span class="p">)</span>
            <span class="n">ncbiId</span><span class="p">,</span><span class="n">symbol</span> <span class="o">=</span> <span class="n">geneIds</span>
            <span class="n">gene2symbol</span><span class="p">[</span><span class="n">ncbiId</span><span class="p">]</span> <span class="o">=</span> <span class="n">symbol</span>
   <span class="n">fid</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
   <span class="k">return</span> <span class="n">gene2symbol</span>
</pre></div>
</div>
<p>Extract the genes into a list</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">pathwayFile</span> <span class="o">=</span> <span class="s">&quot;hsa00860.txt&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pathway</span> <span class="o">=</span> <span class="n">pathwayFile</span><span class="p">[:</span><span class="o">-</span><span class="mi">4</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">geneList</span> <span class="o">=</span> <span class="n">get_genes</span><span class="p">(</span><span class="n">pathway</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">geneList</span><span class="o">.</span><span class="n">items</span><span class="p">()[:</span><span class="mi">3</span><span class="p">])</span>
<span class="go">[(&#39;7363&#39;, &#39;UGT2B4&#39;), (&#39;212&#39;, &#39;ALAS2&#39;), (&#39;210&#39;, &#39;ALAD&#39;)]</span>
</pre></div>
</div>
<p>Specify a location for the analysis files.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">gsaDir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s">&quot;.&quot;</span><span class="p">,</span><span class="s">&quot;gsa-path&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">gsaDir</span><span class="p">):</span>
<span class="gp">&gt;&gt;&gt; </span>   <span class="n">os</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">gsaDir</span><span class="p">)</span>
</pre></div>
</div>
<p>Make imports and specify variables</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">htsint</span> <span class="kn">import</span> <span class="n">GeneOntology</span><span class="p">,</span><span class="n">TermDistances</span><span class="p">,</span><span class="n">GeneDistances</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">useIea</span> <span class="o">=</span> <span class="bp">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aspect</span> <span class="o">=</span> <span class="s">&quot;biological_process&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_aspect</span> <span class="o">=</span> <span class="s">&#39;bp&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">taxaList</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;9606&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">go</span> <span class="o">=</span> <span class="n">GeneOntology</span><span class="p">(</span><span class="n">taxaList</span><span class="p">,</span><span class="n">useIea</span><span class="o">=</span><span class="n">useIea</span><span class="p">,</span><span class="n">aspect</span><span class="o">=</span><span class="n">aspect</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">termsPath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gsaDir</span><span class="p">,</span><span class="s">&quot;go-terms-</span><span class="si">%s</span><span class="s">.pickle&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">_aspect</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">graphPath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gsaDir</span><span class="p">,</span><span class="s">&quot;go-graph-</span><span class="si">%s</span><span class="s">.pickle&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">_aspect</span><span class="p">))</span>
</pre></div>
</div>
<p>By default <tt class="docutils literal"><span class="pre">htsint</span></tt> works at the genome level so to get term distances with respect to a subset of genes we need to specify an <cite>accepted</cite> list.  The rest of the process is the same as was shown in the gsa example.</p>
<p>Create the gene-term dictionaries</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">geneIds</span> <span class="o">=</span> <span class="n">geneList</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">go</span><span class="o">.</span><span class="n">create_dicts</span><span class="p">(</span><span class="n">termsPath</span><span class="p">,</span><span class="n">accepted</span><span class="o">=</span><span class="n">geneIds</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gene2go</span><span class="p">,</span><span class="n">go2gene</span> <span class="o">=</span> <span class="n">go</span><span class="o">.</span><span class="n">load_dicts</span><span class="p">(</span><span class="n">termsPath</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="s">&quot;pathway genes with terms: </span><span class="si">%s</span><span class="s">/</span><span class="si">%s</span><span class="s">&quot;</span><span class="o">%</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gene2go</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span><span class="nb">len</span><span class="p">(</span><span class="n">geneIds</span><span class="p">)))</span>
</pre></div>
</div>
<p>Create the term specific graph</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">G</span> <span class="o">=</span> <span class="n">go</span><span class="o">.</span><span class="n">create_gograph</span><span class="p">(</span><span class="n">termsPath</span><span class="o">=</span><span class="n">termsPath</span><span class="p">,</span><span class="n">graphPath</span><span class="o">=</span><span class="n">graphPath</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="s">&quot;Term graph for with </span><span class="si">%s</span><span class="s"> nodes successfully created.&quot;</span><span class="o">%</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">G</span><span class="o">.</span><span class="n">nodes</span><span class="p">())))</span>
</pre></div>
</div>
<p>Calculate the distances between terms</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">termDistancePath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gsaDir</span><span class="p">,</span><span class="s">&quot;term-distances-</span><span class="si">%s</span><span class="s">.npy&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">_aspect</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">td</span> <span class="o">=</span> <span class="n">TermDistances</span><span class="p">(</span><span class="n">termsPath</span><span class="p">,</span><span class="n">graphPath</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="s">&quot;total distances to evaluate: </span><span class="si">%s</span><span class="s">&quot;</span><span class="o">%</span><span class="n">td</span><span class="o">.</span><span class="n">totalDistances</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">td</span><span class="o">.</span><span class="n">run_with_multiprocessing</span><span class="p">(</span><span class="n">termDistancePath</span><span class="p">,</span><span class="n">cpus</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
</pre></div>
</div>
<p>Map the term distances into gene space</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">geneDistancePath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gsaDir</span><span class="p">,</span><span class="s">&quot;gene-distances-</span><span class="si">%s</span><span class="s">.csv&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">_aspect</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gd</span> <span class="o">=</span> <span class="n">GeneDistances</span><span class="p">(</span><span class="n">termsPath</span><span class="p">,</span><span class="n">termDistancePath</span><span class="p">,</span><span class="n">outFile</span><span class="o">=</span><span class="n">geneDistancePath</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gd</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>
</pre></div>
</div>
<p>Run the parameter search for spectral clustering</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">silvalFile</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="s">&quot;\.csv&quot;</span><span class="p">,</span><span class="s">&quot;-scparams-sv.csv&quot;</span><span class="p">,</span><span class="n">geneDistancePath</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">clustersFile</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="s">&quot;\.csv&quot;</span><span class="p">,</span><span class="s">&quot;-scparams-cl.csv&quot;</span><span class="p">,</span><span class="n">geneDistancePath</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">scps</span> <span class="o">=</span> <span class="n">SpectralClusterParamSearch</span><span class="p">(</span><span class="n">geneDistancePath</span><span class="p">,</span><span class="n">dtype</span><span class="o">=</span><span class="s">&#39;distance&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">scps</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">chunks</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span><span class="n">kRange</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span><span class="mi">11</span><span class="p">))</span>
</pre></div>
</div>
<p>Plot the parameter search</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">psFigureFile</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gsaDir</span><span class="p">,</span><span class="s">&quot;param-scan-</span><span class="si">%s</span><span class="s">.png&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">_aspect</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">scr</span> <span class="o">=</span> <span class="n">SpectralClusterResults</span><span class="p">(</span><span class="n">silvalFile</span><span class="p">,</span><span class="n">clustersFile</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">scr</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">figName</span><span class="o">=</span><span class="n">psFigureFile</span><span class="p">)</span>
</pre></div>
</div>
<div class="align-center figure">
<a class="reference internal image-reference" href="_images/param-scan-bp1.png"><img alt="parameter scan" src="_images/param-scan-bp1.png" style="width: 700.0px; height: 600.0px;" /></a>
</div>
<p>Run spectral clustering with selected parameters</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">k</span> <span class="o">=</span> <span class="mi">3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sigma</span> <span class="o">=</span> <span class="mf">0.43</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">labelsPath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gsaDir</span><span class="p">,</span><span class="s">&quot;sc-labels-</span><span class="si">%s</span><span class="s">.csv&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">_aspect</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span> <span class="o">=</span> <span class="n">SpectralCluster</span><span class="p">(</span><span class="n">geneDistancePath</span><span class="p">,</span><span class="n">dtype</span><span class="o">=</span><span class="s">&#39;distance&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">k</span><span class="p">,</span><span class="n">sk</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span><span class="n">sigma</span><span class="o">=</span><span class="n">sigma</span><span class="p">,</span><span class="n">verbose</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">labelsPath</span><span class="o">=</span><span class="n">labelsPath</span><span class="p">)</span>
</pre></div>
</div>
<p>The genes and their labels are saved in <cite>labelsPath</cite>.</p>
</div>


          </div>
        </div>
      </div>
        </div>
        <div class="sidebar">
          <h3>Table Of Contents</h3>
          <ul>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="database.html">Database</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="database-cookbook.html">Database cookbook</a></li>
<li class="toctree-l1"><a class="reference internal" href="blast.html">Using BLAST</a></li>
<li class="toctree-l1"><a class="reference internal" href="annotation-fetching.html">Gene Ontology annotation fetching</a></li>
<li class="toctree-l1"><a class="reference internal" href="deseq-example.html">Gene level differential expression</a></li>
<li class="toctree-l1"><a class="reference internal" href="gsa-example.html">Gene set analysis example</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="">Pathway example</a></li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References</a></li>
</ul>

          <h3 style="margin-top: 1.5em;">Search</h3>
          <form class="search" action="search.html" method="get">
            <input type="text" name="q" />
            <input type="submit" value="Go" />
            <input type="hidden" name="check_keywords" value="yes" />
            <input type="hidden" name="area" value="default" />
          </form>
          <p class="searchtip" style="font-size: 90%">
            Enter search terms or a module, class or function name.
          </p>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

    <div class="footer-wrapper">
      <div class="footer">
        <div class="left">
          <a href="gsa-example.html" title="Gene set analysis"
             >previous</a> |
          <a href="references.html" title="Works cited"
             >next</a> |
          <a href="genindex.html" title="General Index"
             >index</a>
            <br/>
            <a href="_sources/pathway-example.txt"
               rel="nofollow">Show Source</a>
        </div>

        <div class="right">
          
    <div class="footer">
        &copy; Copyright 2015, A. Richards.
      Last updated on Jul 30, 2015.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.3.
    </div>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

  </body>
</html>